Implement improved & consistent argument parsing

standage opened the pull request:
The sourmash CLI is a bit hairy using a combination of approaches to parse subcommand and subsubcommand arguments. This is my first pass at implementing a consistent approach using vanilla argparse. A lot of copy/paste is still required, and some tests will need to be updated, but the core design I propose is well demonstrated already.

**UPDATE**: The proposed CLI implementation is complete—that is, `--help` works for all commands and subcommands. Having the CLI invoke the appropriate methods is pending. Here I propose making a new subcommand group (e.g. `sourmash sbt combine` instead of `sourmash sbt_combine`, `sourmash sbt categorize` instead of `sourmash categorize`, and so on). This is also implemented, along with a few dirty hacks to support the old CLI (dirty hacks are annotated to simplify removal on a major version bump).

Closes #783.

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

standage commented on cli-sandbox/cli/signature/__init__.py:
```diff
@@ -0,0 +1,29 @@
+from . import merge
+from . import rename
+from . import subtract
+from . import ingest
```
For some reason they won't allow me to create a Python module named `import`! 😉

standage commented on cli-sandbox/README.md:
```diff
@@ -0,0 +1,71 @@
+The `cli/` directory here would be placed in the sourmash root directory.
+
+Note that the actual arguments are incomplete for the examples below, and incorrect for all the commands/subcommands not shown below.
+Fixing all of that will be a tedious copy/paste job.
```
Reiterating: I haven't actually filled in most of the arguments. For the most part, what's in place now are placeholders.

ctb commented:
(do you want contributions to this PR, if I feel so motivated? :)

On Thu, Dec 05, 2019 at 03:50:28PM -0800, Daniel Standage wrote:
> standage commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,71 @@
> +# Proposed CLI design
> +
> +This code demonstrates the mechanics of implementing a CLI with multiple levels of subcommand nesting.
> +The `cli/` directory here would be placed in the sourmash root directory.
> +
> +Note that the actual arguments are incomplete for the examples below, and incorrect for all the commands/subcommands not shown below.
> +Fixing all of that will be a tedious copy/paste job.
> 
> Reiterating: I haven't actually filled in most of the arguments. For the most part, what's in place now are placeholders.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/sourmash/pull/785#pullrequestreview-327926301
-- 
C. Titus Brown, ctbrown@ucdavis.edu

standage commented:
> (do you want contributions to this PR, if I feel so motivated? :)

I'm not expecting any this early, but any feedback or contributions would be welcome. Pointers and contributions would *certainly* be welcome as the CLI matures and we need to start updating the test suite to compensate.

standage commented:
Ok, the CLI is fully implemented—as in `sourmash <cmd> --help` and `sourmash <cmd> <subcmd> --help` displays the correct options for all commands and subcommands. The only (sub)commands that are fully functional at the moment are `sourmash info` and `sourmash signature describe`. But hopefully this is enough to show what I have in mind.

```
$ sourmash
usage: sourmash [-h] [-v] [-q] cmd ...

Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.

Commands:
  Invoke "sourmash <cmd> --help" for more details on executing each command.

  cmd            compare -- compute -- gather -- info -- lca -- plot -- sbt --
                 search -- signature

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
$
$ sourmash info --help
usage: sourmash info [-h] [--verbose]

optional arguments:
  -h, --help  show this help message and exit
  --verbose   report versions of khmer and screed
$
$ sourmash info --verbose
== This is sourmash version 2.3.1.dev2+ge416b98. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

sourmash version 2.3.1.dev2+ge416b98
- loaded from path: /Users/standage/Projects/sourmash/sourmash/cli

khmer version 3.0.0a3
- loaded from path: /Users/standage/anaconda3/envs/sourmashdev/lib/python3.7/site-packages/khmer

screed version 1.0.1
- loaded from path: /Users/standage/anaconda3/envs/sourmashdev/lib/python3.7/site-packages/screed
$
$ sourmash signature --help
usage: sourmash signature [-h] subcmd ...

Subcommands:
  Invoke "sourmash signature <subcmd> --help" for more details on executing
  each subcommand.

  subcmd      describe -- downsample -- export -- extract -- flatten -- ingest
              -- intersect -- merge -- overlap -- rename -- subtract

Options:
  -h, --help  show this help message and exit
$
$ sourmash signature describe --help
usage: sourmash signature describe [-h] [-q] [--csv FILE]
                                   signatures [signatures ...]

positional arguments:
  signatures

optional arguments:
  -h, --help   show this help message and exit
  -q, --quiet  suppress non-error output
  --csv FILE   output information to a CSV file
$
$ sourmash signature describe tests/test-data/2.fa.sig
== This is sourmash version 2.3.1.dev2+ge416b98. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loaded 3 signatures total.
---
signature filename: tests/test-data/2.fa.sig
signature: CP001071.1 Akkermansia muciniphila ATCC BAA-835, complete genome
source file: 2.fa
md5: f372e47893edd349e5956f8b0d8dcbf7
k=21 molecule=DNA num=0 scaled=1000 seed=42 track_abundance=0
size: 2608
signature license: CC0

---
signature filename: tests/test-data/2.fa.sig
signature: CP001071.1 Akkermansia muciniphila ATCC BAA-835, complete genome
source file: 2.fa
md5: f3a90d4e5528864a5bcc8434b0d0c3b1
k=31 molecule=DNA num=0 scaled=1000 seed=42 track_abundance=0
size: 2701
signature license: CC0

---
signature filename: tests/test-data/2.fa.sig
signature: CP001071.1 Akkermansia muciniphila ATCC BAA-835, complete genome
source file: 2.fa
md5: 43f3b48e59443092850964d355a20ac0
k=51 molecule=DNA num=0 scaled=1000 seed=42 track_abundance=0
size: 2585
signature license: CC0
```

## Design

Each command/subcommand will have a:

- `subparser` function for defining its command line interface
- a driver function (with a name that matches the command/subcommand) with the implementation
- a `main` function acting as a thin wrapper around the driver function; facilitates automated testing using simulate command-line arguments

standage commented:
Part of the redesign is to have the SBT subcommands grouped under the `sourmash sbt` command: `sourmash sbt merge`, `sourmash sbt watch`, and so on. Of course, we can't break the API without a major version bump, so I added some dirty hacks to support both the old and new API simultaneously. Each instance is marked as "dirty hack" with a note that, if desired, it can be removed with a major version bump.

Incidentally, `sourmash signature import` was difficult to implement using my strategy since modules cannot be named `import`. So I used a similar dirty hack to make that work as well.

standage commented on sourmash/cli/__init__.py:
```diff
@@ -0,0 +1,112 @@
+            args.subcmd = 'watch'
+        if hasattr(args, 'subcmd') and args.subcmd == 'compare_csv':
+            args.subcmd = 'compare'
+        # END: dirty hacks to simultaneously support new and previous interface
```
Part of the dirty hacks mentioned

standage commented on sourmash/cli/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    sbt.combine.alt_subparser(sub)
+    sbt.index.subparser(sub)
+    sbt.watch.subparser(sub)
+    # END: dirty hacks to simultaneously support new and previous interface
```
Moar dirty hacks: support both `sourmash categorize` and `sourmash sbt categorize`, both `sourmash sbt_combine` and `sourmash sbt combine`, and so on.

luizirber replied:
`categorize` is potentially an [index operation](https://github.com/dib-lab/sourmash/pull/556#issuecomment-449001254), should work with `lca` too (so I suggest keeping it as a upper level, instead of inside `sbt`).

standage replied:
Gotcha. Well, that only leaves the `sbt_combine` as a candidate for grouping sbt-specific subcommands, so I'll just undo that grouping and revert those 4 subcommands to top-level commands.

standage commented on sourmash/cli/lca/compare.py:
```diff
@@ -0,0 +1,56 @@
+        '--no-headers', action='store_true',
+        help='no headers present in taxonomy spreadsheet'
+    )
+    subparser.add_argument('-f', '--force', action='store_true')
```
Another example of the dirty hack scheme: duplicated subparser definitions, presumably to be removed with a major version bump.

standage commented:
**Short version**: intermediate review plz?

**Long version**: Ok, I think I've reached an inflection point. Everything I've done up to this point should be stable for the long term. Minor updates to the CLI on the master branch would have to be recapitulated in this branch, but I expect those types of changes to be infrequent and minimal.

But hooking up the new CLI to the command/subcommand implementations is going to require moving around A LOT of code, and doing so in a way that git will have difficulty tracking. I don't think we should start this process until we are confident we can resolve it quickly. Any other PRs that are merged in parallel in the mean time will have to be recapitulated here, and I'm not enthusiastic about coordinating that.

So before I proceed here, could you folks provide some feedback on what I've done so far? Any questions or concerns?

If y'all support where I'm going with this, I'll be happy to take the lead on the next steps. It'll be tedious, but I don't think it'll take an inordinate amount of time. The last thing I want to do is spend several hours updating the command/subcommand implementations, only to then spend the next few months trying to keep the PR in sync with parallel updates.

standage commented:
By the way, I've copy/pasted some terminal output above, but the best way to see how this is all working might just be to checkout the `cli` branch and start playing around with the `sourmash` command.

luizirber commented:
This is really great, @standage!

I was partially doing this with subparsers in [#245](https://github.com/dib-lab/sourmash/pull/245/files#diff-a83b6a29dae5cdd59d81cc5313fdd6c6R94), but I totally support moving all the argparsing into `sourmash.cli`

I was keeping the [`subparsers`-equivalent function](https://github.com/dib-lab/sourmash/pull/245/files#diff-56040d40281982d44bcd65f91eb0b132R371) near the actual command to help doing proper updates to the function when changing the CLI, and then attaching it to the function (yes, yes, it's a hack) to be able to call it easily in the [`__main__` function](https://github.com/dib-lab/sourmash/pull/245/files#diff-a83b6a29dae5cdd59d81cc5313fdd6c6R99). This is not needed in this PR (since it will still be passing the `args` object around, and not exploding it like I did [here](https://github.com/dib-lab/sourmash/pull/245/files#diff-56040d40281982d44bcd65f91eb0b132R105)), but will make #245 easier to implement.

luizirber commented:
> But hooking up the new CLI to the command/subcommand implementations is going to require moving around A LOT of code, and doing so in a way that git will have difficulty tracking. I don't think we should start this process until we are confident we can resolve it quickly. Any other PRs that are merged in parallel in the mean time will have to be recapitulated here, and I'm not enthusiastic about coordinating that.

Does it need to move code? I think we should keep the command code in `sourmash.commands` for now, and just call it. Like this (for `compare`):

```
diff --git a/sourmash/cli/compare.py b/sourmash/cli/compare.py
index e55b925..eaf93b8 100644--- a/sourmash/cli/compare.py
+++ b/sourmash/cli/compare.py@@ -36,4 +36,4 @@ def subparser(subparsers):
  
 def main(args):-    print(args)
+    return sourmash.commands.compare(args)
```

> If y'all support where I'm going with this, I'll be happy to take the lead on the next steps. It'll be tedious, but I don't think it'll take an inordinate amount of time. The last thing I want to do is spend several hours updating the command/subcommand implementations, only to then spend the next few months trying to keep the PR in sync with parallel updates.

Truer wods were never spoken. Keeping PR in sync with parallel updates is horrible =]

standage commented:
> Does it need to move code? I think we should keep the command code in sourmash.commands for now, and just call it.

That's true. At that point, I was mixing concerns: having a functional CLI, and then being able to invoke that CLI conveniently from Python/the test suite (a la #245).

standage commented:
The suggestion of moving code referred to the design proposed at the end of [this comment](https://github.com/dib-lab/sourmash/pull/785#issuecomment-562444325).

ctb commented:
(note, adjusted title)

luizirber commented:
@standage sorry for all the changes landing on master =(

But... there is a window for including this in `3.0.0-rc2`, and release it together with `3.0.0`. Would you be interested in working on this after the holidays? (and, if not, can I take over the branch?)

standage commented:
> sorry for all the changes landing on master

No worries. As I said earlier, what I've done so far is pretty isolated from the rest of the codebase.

> Would you be interested in working on this after the holidays?

Yep, I had intended to finish this already. But...you know how it goes.

> if not, can I take over the branch?

If I'm ever moving too slowly, feel free to push ahead without me. 😀

luizirber commented:
> > if not, can I take over the branch?
> 
> If I'm ever moving too slowly, feel free to push ahead without me. grinning

Taking over sounded too aggressive, so I opened #811 (merging into this one) for implementations of all commands (pretty much just calling what already exists). There are 16 tests still failing, most because commands moved (like `multigather`)

Some notes:
- duplicated commands (the function that runs after args are parsed): I'm very inclined to keep the ones in `sourmash.cli`, but in the unlikely case there is someone out there calling the `sourmash.commands` variations in their code it is easier to transition than rewriting for the new ones. The ones in `sourmash.cli` are very close to what I did for compute in #245 and I would like to see everything going in this directions... but maybe in sourmash 4.0?
  * describe: in sourmash.cli.sig and sourmash.sig
  * info: in sourmash.cli.info and sourmash.commands
- can't set default for ksize in sourmash.cli.utils.add_ksize_arg. When I first saw https://github.com/dib-lab/sourmash/pull/785/files#diff-e1ba488951dcb791667fc234a1958321R83-R87 I was confused: why the `default` in the function call is not used in the parser? Turns out there are many places in the code that check `if args.ksize is None`, and then depend on that for behavior... I had to [add an override in sourmash.lca.command_index.index](https://github.com/dib-lab/sourmash/pull/811/files#diff-0f55e516e4633cb552e64fa6a352bfcbR330-R331), because it was easier than breaking/fixing all the other places where it failed.

luizirber commented:
Any ideas on how to get nicer usage/help messages? Current ones for `sourmash lca` and `sourmash sig` are manually defined, and harder to keep updated... But they are nicer =]
https://github.com/dib-lab/sourmash/blob/ba5b5a1b630340f3271adf1a6f94a2a1e48e09ee/sourmash/lca/__main__.py#L12L27
https://github.com/dib-lab/sourmash/blob/ba5b5a1b630340f3271adf1a6f94a2a1e48e09ee/sourmash/sig/__main__.py#L18L39

standage commented:
> Any ideas on how to get nicer usage/help messages? Current ones for `sourmash lca` and `sourmash sig` are manually defined, and harder to keep updated... But they are nicer =]

Yeah, it's a trade off: flexibility or convenience. It is possible to have *some* middle ground using [`RawDescriptionHelpFormatter`](https://docs.python.org/3/library/argparse.html#formatter-class), but to the extent that you manually duplicate argparse's functionality you incur the burden of maintenance.

The output of `sourmash -h` is currently as follows. I *do* think we can do better than this.

```
usage: sourmash [-h] [-v] [-q] cmd ...

Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.

Commands:
  Invoke "sourmash <cmd> --help" for more details on executing each command.

  cmd            categorize -- compare -- compute -- dump -- gather --
                 import_csv -- index -- info -- lca -- migrate -- multigather
                 -- plot -- sbt -- sbt_combine -- search -- sig -- storage --
                 watch

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
```

Maybe with something like this, although it does run a bit long. We could easily build it automatically by recursively traversing the `sourmash/cli/` directory, as long as we enforce that the directory and file names match the interface definition.

```
usage: sourmash [-h] [-v] [-q] cmd ...

Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.
Invoke "sourmash <cmd> --help" for more details on executing each command.

    sourmash categorize --help
    sourmash compare --help
    sourmash compute --help
    sourmash dump --help
    sourmash gather --help
    sourmash import_csv --help
    sourmash index --help
    sourmash info --help
    sourmash lca classify --help
    sourmash lca compare --help
    sourmash lca gather --help
    sourmash lca index --help
    sourmash lca rankinfo --help
    sourmash lca summarize --help
    sourmash migrate --help
    sourmash multigather --help
    sourmash plot --help
    sourmash sbt_combine --help
    sourmash search --help
    sourmash sig describe --help
    sourmash sig downsample --help
    sourmash sig export --help
    sourmash sig extract --help
    sourmash sig filter --help
    sourmash sig flatten --help
    sourmash sig ingest --help
    sourmash sig intersect --help
    sourmash sig merge --help
    sourmash sig overlap --help
    sourmash sig rename --help
    sourmash sig subtract --help
    sourmash storage convert -h --help
    sourmash watch --help

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
```

We could potentially even group the commands into sections, reflecting the command/subcommand structure, with "Basic commands" being the default.

```
usage: sourmash [-h] [-v] [-q] cmd ...

Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.
Invoke "sourmash <cmd> --help" for more details on executing each command.

    Basic operations
        sourmash categorize --help
        sourmash compare --help
        sourmash compute --help
        sourmash dump --help
        sourmash gather --help
        sourmash import_csv --help
        sourmash index --help
        sourmash info --help
        sourmash migrate --help
        sourmash multigather --help
        sourmash plot --help
        sourmash sbt_combine --help
        sourmash search --help
        sourmash watch --help

    Lowest common ancestor (LCA) based operations
        sourmash lca classify --help
        sourmash lca compare --help
        sourmash lca gather --help
        sourmash lca index --help
        sourmash lca rankinfo --help
        sourmash lca summarize --help
    
    Operations on signatures (MinHash sketches)
        sourmash sig describe --help
        sourmash sig downsample --help
        sourmash sig export --help
        sourmash sig extract --help
        sourmash sig filter --help
        sourmash sig flatten --help
        sourmash sig ingest --help
        sourmash sig intersect --help
        sourmash sig merge --help
        sourmash sig overlap --help
        sourmash sig rename --help
        sourmash sig subtract --help
    
    Operations on storage
        sourmash storage convert -h --help

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
``` 

By the way, are additional `sourmash storage` subcommands planned? At the moment it appears that `convert` is the only one implemented.

luizirber commented:
> Yeah, it's a trade off: flexibility or convenience. It is possible to have _some_ middle ground using [`RawDescriptionHelpFormatter`](https://docs.python.org/3/library/argparse.html#formatter-class), but to the extent that you manually duplicate argparse's functionality you incur the burden of maintenance.

I started doing one that digged into the argparse internals and... it got ugly fast. =/

> We could potentially even group the commands into sections, reflecting the command/subcommand structure, with "Basic commands" being the default.

Ooooh, this one is NICE!

> By the way, are additional `sourmash storage` subcommands planned? At the moment it appears that `convert` is the only one implemented.

Yes, I'll be more mindful when adding commands in the future... for now it's better to keep what exist working, and then remove on 4.0 if no other commands show up.

standage commented:
Ok, `sourmash --help` now produces this.

```
usage: sourmash [-h] [-v] [-q]

Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.

    Basic operations
        sourmash categorize --help
        sourmash compare --help
        sourmash compute --help
        sourmash dump --help
        sourmash gather --help
        sourmash import_csv --help
        sourmash index --help
        sourmash info --help
        sourmash migrate --help
        sourmash multigather --help
        sourmash plot --help
        sourmash sbt_combine --help
        sourmash search --help
        sourmash watch --help

    Lowest common ancestor (LCA) based operations
        sourmash lca classify --help
        sourmash lca compare --help
        sourmash lca gather --help
        sourmash lca index --help
        sourmash lca rankinfo --help
        sourmash lca summarize --help

    Operations on signatures (MinHash sketches)
        sourmash sig describe --help
        sourmash sig downsample --help
        sourmash sig export --help
        sourmash sig extract --help
        sourmash sig filter --help
        sourmash sig flatten --help
        sourmash sig ingest --help
        sourmash sig intersect --help
        sourmash sig merge --help
        sourmash sig overlap --help
        sourmash sig rename --help
        sourmash sig subtract --help

    Operations on storage
        sourmash storage convert --help

Instructions:
  Invoke "sourmash <cmd> --help" or "sourmash <cmd> <subcmd> --help"
  for more details on executing each command.

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
```

Command groups have a similar summary. `sourmash lca --help` produces this.

```
usage: sourmash lca [-h]

Subcommands:
  Invoke "sourmash lca <subcmd> --help" for more details on executing each subcommand.
  
      Operations
          sourmash lca classify --help
          sourmash lca compare --help
          sourmash lca gather --help
          sourmash lca index --help
          sourmash lca rankinfo --help
          sourmash lca summarize --help

Options:
  -h, --help  show this help message and exit
```

And finally, each operation has its own usage. `sourmash lca index --help` produces this.

```
usage: sourmash lca index [-h] [--scaled S] [-k K] [-q] [-d] [-C C] [--tabs]
                          [--no-headers] [--split-identifiers] [-f]
                          [--traverse-directory] [--report REPORT]
                          [--require-taxonomy]
                          csv lca_db_out signatures [signatures ...]

positional arguments:
  csv                   taxonomy spreadsheet
  lca_db_out            output database name
  signatures            one or more sourmash signatures

optional arguments:
  -h, --help            show this help message and exit
  --scaled S
  -k K, --ksize K       k-mer size; default=31
  -q, --quiet           suppress non-error output
  -d, --debug           output debugging output
  -C C, --start-column C
                        column at which taxonomic assignments start; default=2
  --tabs                input spreadsheet is tab-delimited; default is commas
  --no-headers          no headers present in taxonomy spreadsheet
  --split-identifiers   split names in signatures on whitspace and period
  -f, --force
  --traverse-directory  load all signatures underneath directories
  --report REPORT       output a report on anomalies, if any
  --require-taxonomy    ignore signatures with no taxonomy entry
```

ctb commented:
is there a way to hide certain operations, and/or prioritize others in the
output?

e.g. compare, compute, search, gather should be in the basic operations,
and maybe index, but not the others.

lca could all be grouped under one, same for sig.

I have a few goals here --

* limit complexity for new users
* keep screen display for basic help < 25 rows

we can probably keep a custom help message here, come to think of it.

ctb commented:
p.s. and wow, this is super cool :)

luizirber commented:
>I have a few goals here -- 
>* limit complexity for new users
Fair point

>* keep screen display for basic help < 25 rows 
>we can probably keep a custom help message here, come to think of it.

But for these two... a custom help message always has the risk of becoming outdated (when a new command is added, but message is not updated), and showing all the subcommands also keep us mindful of how complex the CLI is becoming =]

Maybe keep only the basic commands, and for lca/sig/storage point to the subcommand help message (instead of listing everything)?

luizirber commented:
> p.s. and wow, this is super cool :)

super cool indeed, awesome job @standage!

ctb commented:
On Thu, Jan 02, 2020 at 12:37:32PM -0800, Luiz Irber wrote:
> 
> >I have a few goals here -- 
> >* limit complexity for new users
> Fair point
> 
> >* keep screen display for basic help < 25 rows 
> >we can probably keep a custom help message here, come to think of it.
> 
> But for these two... a custom help message always has the risk of becoming outdated (when a new command is added, but message is not updated), and showing all the subcommands also keep us mindful of how complex the CLI is becoming =]
> 
> Maybe keep only the basic commands, and for lca/sig/storage point to the subcommand help message (instead of listing everything)?

yes, that's what I'm thinking

standage commented:
Well this gets us to 32 rows...30 if we sacrifice the top two lines. Not quite 25, but better than before.

```
Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.

Usage instructions:
    Basic operations
        sourmash categorize --help
        sourmash compare_csv --help
        sourmash compute --help
        sourmash dump --help
        sourmash gather --help
        sourmash import_csv --help
        sourmash index --help
        sourmash info --help
        sourmash migrate --help
        sourmash multigather --help
        sourmash plot --help
        sourmash sbt_combine --help
        sourmash search --help
        sourmash watch --help

    Lowest common ancestor (LCA) based operations
        sourmash lca --help

    Operations on signatures (MinHash sketches)
        sourmash sig --help

    Operations on storage
        sourmash storage --help

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
```

And the subcommand group help.

```
Lowest common ancestor (LCA) utilities:
  Operations
      sourmash lca classify --help
      sourmash lca compare --help
      sourmash lca gather --help
      sourmash lca index --help
      sourmash lca rankinfo --help
      sourmash lca summarize --help

Options:
  -h, --help  show this help message and exit
```

ctb commented:
I think categorize, dump, import_csv, migrate, sbt_combine, and watch
are all expert commands that don't belong in there.

standage commented:
> I think categorize, dump, import_csv, migrate, sbt_combine, and watch are all expert commands that don't belong in there.

Ok, you command elitist! 😀 Now it looks like this.

```
Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.

Usage instructions:
    Basic operations
        sourmash compare_csv --help
        sourmash compute --help
        sourmash gather --help
        sourmash index --help
        sourmash info --help
        sourmash multigather --help
        sourmash plot --help
        sourmash search --help

    Lowest common ancestor (LCA) based operations
        sourmash lca --help

    Operations on signatures (MinHash sketches)
        sourmash sig --help

    Operations on storage
        sourmash storage --help

Options:
  -h, --help     show this help message and exit
  -v, --version  show program's version number and exit
  -q, --quiet    don't print citation information
```

You'll just have to maintain [this catalog of elitist commands](https://github.com/dib-lab/sourmash/pull/785/commits/6aa62598dbea49c9ff9e75a8fe8469a913417bc8#diff-e1ba488951dcb791667fc234a1958321R81). If you know the secrets of sourmash, you can still get the usage info.

```
$ sourmash dump -h 
usage:  dump [-h] [-k K] filenames [filenames ...]

positional arguments:
  filenames

optional arguments:
  -h, --help       show this help message and exit
  -k K, --ksize K  k-mer size; default=31
$ sourmash migrate -h 
usage:  migrate [-h] sbt_name

positional arguments:
  sbt_name    name to save SBT into

optional arguments:
  -h, --help  show this help message and exit
$
```

luizirber commented:
I would also take out `multigather` (should be covered by `gather` now? and removed in 4.0?) and... is `compare_csv` equivalent to `compare`? I think we should have the later, and not the former...

> ```
> Compute, compare, manipulate, and analyze MinHash sketches of DNA sequences.
> 
> Usage instructions:
>     Basic operations
>         sourmash compare_csv --help
>         sourmash compute --help
>         sourmash gather --help
>         sourmash index --help
>         sourmash info --help
>         sourmash multigather --help
>         sourmash plot --help
>         sourmash search --help
> 
>     Lowest common ancestor (LCA) based operations
>         sourmash lca --help
> 
>     Operations on signatures (MinHash sketches)
>         sourmash sig --help
> 
>     Operations on storage
>         sourmash storage --help

luizirber commented:
(and sorry for all the bikeshedding...)

standage commented:
> I would also take out multigather (should be covered by gather now? and removed in 4.0?)

Done

> and... is compare_csv equivalent to compare? I think we should have the later, and not the former...

Oops, my fault. I thought I had changed it to `compare` a few weeks ago, and that changing it to `compare_csv` was restoring the original name. I've changed it back to `compare`.

luizirber commented:
> Oops, my fault. I thought I had changed it to `compare` a few weeks ago, and that changing it to `compare_csv` was restoring the original name. I've changed it back to `compare`.

Both exist, `compare` is a basic command and `compare_csv` is part of the `lca` subcommand (that's why there are 3 tests failing now =P)

standage commented on tests/test_sourmash.py:
```diff
@@ -43,9 +44,8 @@ def test_run_sourmash_subcmd_help():
-    # assert error was printed
-    assert any("invalid choice: 'sbt'" in o for o in (err, out))
+    assert "invalid choice:" in err
+    assert "'sbt' (choose from" in err
```
Nice that you don't have to do the stdout/stderr dance anymore.

standage commented:
I think this is ready for review and merge.

At some points I flirted with changing the interface, but I've tried to roll back all those changes. It's worth a careful check though. The only exception is `sourmash sig import`, which is actually implemented underneath as `sourmash sig ingest` (since you can't `from . import import`), but it can be called as `sourmash sig import`.

luizirber commented:
I'll let @ctb have a final look with fresh eyes, but otherwise :100:

ctb commented:
@standage, this looks fantastic! A few comments for you and @luizirber to consider --

* could you add citation information in the base `sourmash` help output? Something like: "sourmash is scientific software. Please cite doi:10.21105/joss.00027."
* under 'basic operations', it's not clear what these commands do. Can we either (a) link to a tutorial or (b) include brief descriptions (e.g. "compare genomes", "create genome signatures", etc.)
* "Lowest common ancestor operations" should be something like "taxonomy operations" I think.
* "Operations on signatures" -> "Manipulate signature files"

(Obviously the command line interface needs some rationalization etc for 4.0, but never mind that for now... :)

As long as I'm in "reviewer" mode, I think the `cli/__init__.py` and other __init__.py files should have a docstring explaining what those modules contain. Nothing long, just ...not empty :).

I'm surprised by use of absolute imports, e.g. in `sourmash/cli/compute.py`, at bottom, `from sourmash.command_compute import compute`. Is this for a reason or? (I don't think it needs to be changed, necessarily, just would like to know if there's a rationale.)

I'm not a huge fan of deeply nested directory structures, FWIW; this:
```
sourmash
|-- cli
|   |-- lca
|   |-- sig
|   `-- storage
|-- lca
`-- sig
```
makes it challenging for less experienced people to find things. Maybe we could consider using prefixes instead, e.g. `cli/lca_*.py` is for the LCA commands. Not for this PR tho.

Let's create an issue to add docstrings to functions in `sourmash/commands.py`. I think this is related to the "let's make sourmash more usable from Python" issue, #586 and similar. (Not for this PR, tho.)

(I'm also an even bigger fan now of making an `sbt` submodule, #177. wrt deeply nested directories - all things in moderation I guess? :)

I'm definitely a fan of merging this now, or soon, and then iterating on the things above - just pls create issues with whatever  from above you decide NOT to do in this PR.

luizirber commented:
> I'm surprised by use of absolute imports, e.g. in `sourmash/cli/compute.py`, at bottom, `from sourmash.command_compute import compute`. Is this for a reason or? (I don't think it needs to be changed, necessarily, just would like to know if there's a rationale.)

Oh, this one was mine. Relative imports here would be pretty weird, no? `from ..command_compute import compute` works, but I though it was clearer to use `sourmash.` instead of `..`.

> I'm not a huge fan of deeply nested directory structures, FWIW; this:
> 
> ```
> sourmash
> |-- cli
> |   |-- lca
> |   |-- sig
> |   `-- storage
> |-- lca
> `-- sig
> ```

Yeah, it's kind of weird to have this "duplicated subtree" under `cli/`, but... It's also a result of mixing the CLI and the Python API in the same place. Otherwise we end in the situation of `commands.py`, which is gigantic and does too much in one place...

> (I'm also an even bigger fan now of making an `sbt` submodule, #177. wrt deeply nested directories - all things in moderation I guess? :)

And complementing the point above, there is also the issue of having different styles/levels of maturity across the codebase. That's how we end with `commands.py`, `sbt*.py`, `lca/`...  But agree on making the SBT submodule.

(But I was also chastised from changing internal code locations in https://github.com/dib-lab/sourmash/pull/812#issuecomment-570920683, so I will just stay quiet for now =P)

ctb commented:
On Sun, Jan 05, 2020 at 07:32:16PM -0800, Luiz Irber wrote:
> > I'm surprised by use of absolute imports, e.g. in `sourmash/cli/compute.py`, at bottom, `from sourmash.command_compute import compute`. Is this for a reason or? (I don't think it needs to be changed, necessarily, just would like to know if there's a rationale.)
> 
> Oh, this one was mine. Relative imports here would be pretty weird, no? `from ..command_compute import compute` works, but I though it was clearer to use `sourmash.` instead of `..`.

ok. there are no downsides then?

> > I'm not a huge fan of deeply nested directory structures, FWIW; this:
> > 
> > ```
> > sourmash
> > |-- cli
> > |   |-- lca
> > |   |-- sig
> > |   `-- storage
> > |-- lca
> > `-- sig
> > ```
> 
> Yeah, it's kind of weird to have this "duplicated subtree" under `cli/`, but... It's also a result of mixing the CLI and the Python API in the same place. Otherwise we end in the situation of `commands.py`, which is gigantic and does too much in one place...

'struth

> > (I'm also an even bigger fan now of making an `sbt` submodule, #177. wrt deeply nested directories - all things in moderation I guess? :)
> 
> And complementing the point above, there is also the issue of having different styles/levels of maturity across the codebase. That's how we end with `commands.py`, `sbt*.py`, `lca/`...  But agree on making the SBT submodule.

it's aliiiiiiiiive

> (But I was also chastised from changing internal code locations in https://github.com/dib-lab/sourmash/pull/812#issuecomment-570920683, so I will just stay quiet for now =P)

hmm, "let's make a new package" is quite different from "I don't like a/b/c/d/e
subdirectories" :)

standage commented:
Okey doke, new commits.

- Overrode the `SourmashParser` class to print the citation when `print_help()` is invoked.
- Added docstrings to each CLI (sub)module; these can be accessed programmatically to provide descriptions of the operations in the help messages; I did this not only for the Basic Operations in `sourmash --help` but also for the subsubcommands in `sourmash lca`, `sourmash sig`, and `sourmash storage`. NOTE: a couple docstrings are only placeholders—**these should be fixed before this PR is merged.**
- Fixed help section heading's per @ctb's suggestions.

Opening separate issues now for the remaining stuff.

standage commented:
- [x] document need for function docstrings in `sourmash/commands.py`; see #828 
- [x] document suggestion to investigate filenames instead of directories for defining CLI nesting structure; see #827
- [ ] fill in docstrings for `sourmash/cli/categorize.py`, `sourmash/cli/dump.py`, `sourmash/cli/import_csv.py`, `sourmash/cli/migrate.py`, `sourmash/cli/multigather.py`, `sourmash/cli/storage/convert.py`

This last one needs someone who knows what they're talking about! 😀

standage commented:
@luizirber and @ctb, I leave this PR in your capable hands.

ctb commented:
docstrings added!

One minor annoyance is that the citation info is printed out twice when you do a module-level subcommand, e.g. `sourmash lca`. If it's an easy fix, that'd be a good 'un. It doesn't happen at the top level (`sourmash`) nor does it happen for an actual command (`sourmash lca search --help`)

```
sourmash lca 

== This is sourmash version 2.2.1.dev8+ga33a904. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

Taxonomic utilities:
  Operations
          sourmash lca classify --help      classify genomes
          sourmash lca compare_csv --help   compare spreadsheets
          sourmash lca gather --help        classify metagenomes
          sourmash lca index --help         create LCA database
          sourmash lca rankinfo --help      database rank info
          sourmash lca summarize --help     summarize mixture

Options:
  -h, --help  show this help message and exit

== This is sourmash version 2.2.1.dev8+ga33a904. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==
```

luizirber commented on sourmash/cli/__init__.py:
```diff
@@ -0,0 +1,125 @@
+        self.citation = citation
+
+    @classmethod
+    def print_citation(cls):
```
I made `print_citation` a class method, and storing the `_citation_printed` at the class level let's us assert that it is only printed once along all commands (because all commands are subclasses of `SourmashParser`).

(the `classmethod` trick is straight out of [snakemake], btw)

[snakemake]: https://github.com/snakemake/snakemake/blob/e4b1e9a41ddf0ed260e3048586ca488b7fa4a148/snakemake/shell.py

luizirber commented:
:tada: Thanks @standage :100:

ctb commented:
yay w00t thanks!!

standage commented:
Awesome. Thanks for your feedback!
